spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkFilters.java

```diff
@@ -161,10 +162,13 @@ public static Expression convert(Filter filter) {
 
         case IN:
           In inFilter = (In) filter;
+          if (Stream.of(inFilter.values()).anyMatch(Objects::isNull)) {
```

Comment:
Why is the previous logic wrong? 

```
col1 IN (1, 2, NULL) -> gives true if col1 is 1
col2 IN (1, 2, NULL) -> gives null if col1 is 10 and null means no match
col3 IN (1, 2, NULL) -> gives null if col1 is null and null means no match
```

Nulls in predicates are handled in a special way as they are same as false. 
There is `ReplaceNullWithFalseInPredicate` in Spark that I worked on 4-5 years ago.

Reply:
We previously removed nulls as Iceberg filters behave as Java filters while Spark passes SQL filters.

Reply:
I am sorry missed the initial discussion. @rdblue @ConeyLiu, could you elaborate?

Reply:
I thought we handled such cases in a special way. For instance, there is `hasNoInFilter` used in the negation to recursively check for nested NOT IN inside NOT. We do handle IN and NOT IN differently.

My worry is that filters like `IN (1, 2, NULL)`, which are perfectly fine to push down, will no longer be pushed down, causing silent performance issues when predicates are computes programatically.

Reply:
Ah, I missed that we have specific handling to ensure that there is no IN nested within NOT. I think you're right that this is okay after all. We should probably double check more cases, like that the handling for `NOT(IN(1, 2))` is also correct for `NOT(IN(1, 2, null)`.

Reply:
Thanks @aokolnychyi @rdblue for the explanations. So the reason that we can push down safety is:
1. `NULL` in a predicate will be optimized into false in Spark.
2. We have special processing for `NOT IN` and don't allow other `IN` nested within `NOT`. The Spark `col NOT IN (1, 2)` is converted to Iceberg expressions: `notNull(col) && notIn(col, 1, 2)`.

I have a question about `IN`. Is it safe to convert `col IN (1, 2)` to Iceberg: `notNull(col) && in(col, 1, 2)` as well? Why not do that?